File: python/UVA/3n+1.py
#!/usr/bin/python

import sys

def sequence(n, cyclecount):
	if n == 1:
		return cyclecount + 1

	if n & 1 == 1:
		currCnt = sequence((3 * n + 1), cyclecount+1)
	else:
		currCnt = sequence(n / 2, cyclecount+1)

	return currCnt

def findMaxCycleCount(i, j):
	maxCnt = 0
	for x in range(i, j+1):
		currCnt = sequence(x, 0)
		if (currCnt > maxCnt):
			maxCnt = currCnt

	return maxCnt

inputList = []
results = []
for line in sys.stdin.readlines():
	strLine = line[:-1].split(" ")
	i, j = int(strLine[0]), int(strLine[1])
	results.append([i,j,findMaxCycleCount(i,j)])

for result in results:
	print result[0], result[1], result[2]

File: c/uva/3n+1.c
#include <stdio.h>
#include <stdlib.h>

int sequence_3n1(int n)
{
	if ( n == 1 )
		return 1;
	if ( n & 1)
		return sequence_3n1(3*n+1) + 1;
	else
		return sequence_3n1(n/2) + 1;
}

int findMaxCycleCount(int i, int j)
{
	int maxCnt = 0;
	int currCnt;

	while( i <= j ) {
		currCnt = sequence_3n1(i);
		if (currCnt > maxCnt)
			maxCnt = currCnt;
		i++;
	}

	return maxCnt;
}

int main()
{
	int m, n;

	while (scanf("%d %d", &m, &n) == 2) {
		if ( m < n) 
			printf("%d %d %d\n", m, n, findMaxCycleCount(m, n));
		else
			printf("%d %d %d\n", m, n, findMaxCycleCount(n, m));
	}
		
	return 0;
}
